### PR TITLE
feat(portfolio): Merge multichain tokens

### PR DESCRIPTION
**Pull request by th0rOdinson**
https://github.com/Balmy-protocol/dca-fe/assets/144699868/78051b52-af54-458d-8fb7-b11604bffac5

### CODE REVIEW

**Comment by 0xged** [comment]
image
Any chance we add the shadow on the network icons? I think it adds a little bit more depth to it. Looks INSANEEEEEEE IM SO EXCITED LFGGG

File: apps/root/src/pages/home/components/portfolio/index.tsx [hunk 1]
```diff
@@ -276,11 +296,59 @@ const Portfolio = ({ selectedWalletOption }: PortfolioProps) => {
...
+    });
+
     const mappedBalances = map(Object.entries(tokenBalances), ([key, value]) => ({
       ...value,
       key,
       relativeBalance:
         assetsTotalValue.wallet && value.balanceUsd ? (value.balanceUsd / assetsTotalValue.wallet) * 100 : 0,
+      price:
+        value.tokens.length > 1
+          ? meanBy(
```
Comment (FiboApe) [inline apps/root/src/pages/home/components/portfolio/index.tsx:347]: wouldnt it be the same if we just send 1?

File: apps/root/src/pages/home/components/portfolio/index.tsx [hunk 2]
```diff
@@ -276,11 +296,59 @@ const Portfolio = ({ selectedWalletOption }: PortfolioProps) => {
...
+          tokenBalances[balanceItemKey] = balanceItem;
+          delete tokenBalances[`${chainId}-${address}`];
+        }
+      });
+
+      if (balanceItem.tokens.length > 1) {
+        // All equivalent tokens should have same decimals, but we check just in case against the first token
+        const totalBalance = balanceItem.tokens.reduce(
+          (acc, { balance, token: equivalentToken }) =>
+            acc + parseUnits(balance.toString(), firstAddedToken.decimals - equivalentToken.decimals) || 0n,
```
Comment (FiboApe) [inline apps/root/src/pages/home/components/portfolio/index.tsx:327]: wait why are we doing this based on the difference of decimals? Shouldn't each balance be calculated by the tokens decimals of that balance?

File: apps/root/src/pages/home/components/portfolio/index.tsx [hunk 3]
```diff
@@ -276,11 +296,59 @@ const Portfolio = ({ selectedWalletOption }: PortfolioProps) => {
...
+          0n
+        );
+        const totalBalanceUsd = balanceItem.tokens.reduce((acc, { balanceUsd }) => acc + (balanceUsd || 0), 0);
+
+        balanceItem.balance = totalBalance;
+        balanceItem.balanceUsd = totalBalanceUsd;
+      } else {
+        balanceItem.balance = balanceItem.tokens[0].balance;
```
Comment (FiboApe) [inline apps/root/src/pages/home/components/portfolio/index.tsx:335]: could we maybe make it generic so we don't have all of this tokens[0]?

File: apps/root/src/state/token-lists/actions.ts [hunk 4]
```diff
@@ -44,7 +44,6 @@ export const fetchTokenDetails = createAppAsyncThunk<Token, { tokenAddress: stri
     const curatedTokenList = parseTokenList({
       tokensLists: {
         ...tokensLists,
-        'custom-tokens': customTokens,
```
[the remark this answers is not in the document]
Comment (th0rOdinson) [reply apps/root/src/state/token-lists/actions.ts:47]: Got it, sending it here as it is wasn't doing anything. So I'll undo this and add it to `curatedLists`

File: apps/root/src/state/token-lists/reducer.ts [hunk 5]
```diff
@@ -16,14 +17,17 @@ export interface TokenListsState {
...
 export const getDefaultByUrl = () => ({
   /* -------------------------------------------------------------------------- */
-  /*                                   General                                  */
+  /*                                   Complete                                 */
   /* -------------------------------------------------------------------------- */
-  'https://raw.githubusercontent.com/balmy-protocol/token-lister/main/token-list-complete.json': {
+  [`https://raw.githubusercontent.com/balmy-protocol/token-lister/${TOKEN_LISTER_BRANCH}/token-list-complete.json`]: {
     name: 'Mean Finance be',
     logoURI: '',
     timestamp: new Date().getTime(),
```
[the remark this answers is not in the document]
Comment (th0rOdinson) [reply apps/root/src/state/token-lists/reducer.ts:33]: It does, i'll think of a better name 😄

File: apps/root/src/pages/home/components/portfolio/index.tsx (hunk 2 above)
Comment (th0rOdinson) [reply apps/root/src/pages/home/components/portfolio/index.tsx:327]: You're right, new approach in progress!

File: apps/root/src/pages/home/components/portfolio/index.tsx [hunk 6]
```diff
@@ -320,38 +326,30 @@ const Portfolio = ({ selectedWalletOption }: PortfolioProps) => {
...
-      } else {
-        balanceItem.balance = balanceItem.tokens[0].balance;
-        balanceItem.balanceUsd = balanceItem.tokens[0].balanceUsd;
-      }
+      const totalBalanceInUnits = balanceItem.tokens.reduce((acc, { balance, token: equivalentToken }) => {
+        return acc + Number(formatUnits(balance, equivalentToken.decimals));
+      }, 0);
+      const totalBalanceUsd = balanceItem.tokens.reduce((acc, { balanceUsd }) => acc + (balanceUsd || 0), 0);
+
+      const totalBalanceInUnitsFormatted = formatUnits(parseUnits(totalBalanceInUnits.toFixed(18), 18), 18);
```
[the remark this answers is not in the document]
Comment (th0rOdinson) [reply apps/root/src/pages/home/components/portfolio/index.tsx:334]: The path I took was:

<ins>For calculating totalAmount (in units):</ins>
1. Parse each token amount into units (by it's own decimal), and accumulate them (type: number)
2. Format it to string, avoiding exponential notation (formatUnits( , 18) + pareUnits( , 18) + toFixed(18))

<ins>For formatting calculated amountInUnits:</ins>
1. In the component return statement, parse it by X decimals, so formatCurrencyAmount can format it by same X decimals. Using this function we can localize, use sigFigs, etc

**Review by FiboApe — APPROVED** [review]

**Pull request merged**